### PR TITLE
Add check alias

### PR DESCRIPTION
**Pull request by rgrinberg**
This alias corresponds to the minimum that is required to build to get tooling
support from merlin and other cmt using tools.

As discussed in #1221. This alias is usually much faster to build than `@install`. It is the alias one would like to build in the background when doing development in the editor.

Initially, I tried implementing this alias by carefully adding the targets when they were created in `Module_compilation` in ml. But it was kind of tedious and error prone, and also required to do some contortions with `on_load_dir` as Module_compilation would only load the rules in the object dirs themselves. It seemed simpler to just rely on the new and flexible globs for this.

### CODE REVIEW

**Comment by rgrinberg** [comment]
That makes sense. I've pushed a commit that limits the alias to the merlin workspace. As a corollary, I wonder if it makes sense to disable cmt/cmti's for the other workspaces.

**Comment by jberdine** [comment]
Thanks! Works really well.

**Comment by rgrinberg** [comment]
@jberdine do you have numbers to report on how much faster is `@check` compared to `@install` (or w/e alias you use for development)

**Comment by jberdine** [comment]
So far only for a small project: `@check` is about 40% faster than building bytecode for one context, and 10x faster than a normal build (which involves 2 contexts, one standard native code, and one flambda -O3). But I expect the gap to widen with the project size. This is with a 4 core / 8 thread machine FWIW.

**Comment by ejgallego** [comment]
> than building bytecode for one context

Is it possible to limit a context to do a byte-only build?

**Comment by jberdine** [comment]
>> than building bytecode for one context
>
> Is it possible to limit a context to do a byte-only build?

I don't know. I explicitly list the toplevel bytecode targets, e.g.: 
`dune build _build/dev/src/foo.bc _build/dev/src/bar.bc`.

**Comment by rgrinberg** [comment]
`@check` shouldn't build more than just bytecode btw, but I'll double check.

> Is it possible to limit a context to do a byte-only build?

No, but I think that if we had https://github.com/ocaml/dune/issues/1073 then it would make it possible. You'd simply set `(modes byte)` in your env for that context.

**Comment by ejgallego** [comment]
> No, but I think that if we had #1073 then it would make it possible. You'd simply set `(modes byte)` in your env for that context.

Great, that is exactly what I'd need to replicate the `make byte` target of Coq. Not that I think the target is very useful but some devs ask for it.

**Comment by rgrinberg** [comment]
> Great, that is exactly what I'd need to replicate the make byte target of Coq. Not that I think the target is very useful but some devs ask for it.

If they'd like a target to quickly use in development then a combination of `@check` and `coqc.bc` should work for them

File: src/lib_modules.ml [hunk 1]
```diff
@@ -159,6 +159,9 @@ let for_compilation t =
   | None, false -> t.modules
   | Some alias, false -> Module.Name_map.add t.modules alias
 
+let has_private_modules t =
+  Module.Name.Map.exists t.modules ~f:(fun m -> not (Module.is_public m))
```
Comment (emillon) [inline src/lib_modules.ml]: What do you think about extracting `Module.is_private` and eta reducing this?

Comment (rgrinberg) [reply src/lib_modules.ml]: seems reasonable

**Comment by rgrinberg** [comment]
I will add documentation once #1409 is merged.

File: src/check_rules.ml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+open Stdune
+
+let exts = [".cmt"; ".cmti"; ".cmi"]
+
+let dev_files p =
+  List.exists ~f:(fun suffix -> Path.is_suffix p ~suffix) exts
```
Comment (diml) [inline src/check_rules.ml]: Small nitpick: it seems more direct to me to do `match Path.extension p with | ".cmt" | ... -> true | _ -> false`. It's faster as well as the compiler optimizes pattern matching on strings.

**Review by ghost — APPROVED** [review]

File: src/check_rules.ml (hunk 2 above)
Comment (rgrinberg) [reply src/check_rules.ml]: Okay, will change it to your suggestion. I actually that my version would have the edge as it wouldn't allocate anything if the closure for `~f` was eliminated.

Comment (diml) [reply src/check_rules.ml]: It's true that you skip the allocation and copying

**Pull request merged**